### PR TITLE
[dynamo] Fix name collision bug for captured cells and locals

### PR DESCRIPTION
**Pull request by StrongerXi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140155
* #140154
* #140153
* #140435
* #140436
* #140152
* #140330
* #140202
* #140201
* #140151
* #140150
* #140149
* __->__ #140036
* #140035

The `export_freevars` method was introduced very early on, for
propagating writes to unboxed cells from child to parent frame, see
https://github.com/pytorch/torchdynamo/commit/d0c10341.

However, it's no longer needed after we started to modify root tracer's
`symbolic_locals` directly for the unboxed cells, see
https://github.com/pytorch/torchdynamo/commit/663e4d92.

As a result, we no longer need `export_freevars`. In fact, it can cause
a very subtle bug when name collision happens across the parent and
child frames during inlining, because the parent frame isn't necessarily
the frame that defined the cell captured by child frame.

In summary, this patch removes the `export_freevars` bits, and adds a
regression test.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Review by williamwen42 — APPROVED** [review]

**Comment by StrongerXi** [comment]
Rebase

**Review by jansel — APPROVED** [review]

**Comment by StrongerXi** [comment]
Rebase

**Comment by StrongerXi** [comment]
Rebase
